File: Source/AlphaTab/Importer/GpxImporter.cs
using AlphaTab.Model;
using AlphaTab.Util;

namespace AlphaTab.Importer
{
    /// <summary>
    /// This ScoreImporter can read Guitar Pro 6 (gpx) files.
    /// </summary>
    internal class GpxImporter : ScoreImporter
    {
        public override string Name => "Guitar Pro 6";

        public override Score ReadScore()
        {
            // at first we need to load the binary file system
            // from the GPX container
            Logger.Info(Name, "Loading GPX filesystem");
            var fileSystem = new GpxFileSystem();
            fileSystem.FileFilter = s =>
                    s.EndsWith(GpxFileSystem.ScoreGpif)
                    || s.EndsWith(GpxFileSystem.BinaryStylesheet)
                    || s.EndsWith(GpxFileSystem.PartConfiguration)
                ;
            fileSystem.Load(Data);
            Logger.Info(Name, "GPX filesystem loaded");

            string xml = null;
            byte[] binaryStylesheet = null;
            byte[] partConfiguration = null;
            foreach (var entry in fileSystem.Files)
            {
                switch (entry.FileName)
                {
                    case GpxFileSystem.ScoreGpif:
                        xml = Platform.Platform.ToString(entry.Data, Settings.Importer.Encoding);
                        break;
                    case GpxFileSystem.BinaryStylesheet:
                        binaryStylesheet = entry.Data;
                        break;
                    case GpxFileSystem.PartConfiguration:
                        partConfiguration = entry.Data;
                        break;
                }
            }

            // lets set the fileSystem to null, maybe the garbage collector will come along
            // and kick the fileSystem binary data before we finish parsing
            fileSystem.Files = null;
            fileSystem = null;

            // the score.gpif file within this filesystem stores
            // the score information as XML we need to parse.
            Logger.Info(Name, "Start Parsing score.gpif");
            var gpifParser = new GpifParser();
            gpifParser.ParseXml(xml, Settings);
            Logger.Info(Name, "score.gpif parsed");

            var score = gpifParser.Score;

            if (binaryStylesheet != null)
            {
                Logger.Info(Name, "Start Parsing BinaryStylesheet");
                var stylesheetParser = new BinaryStylesheetParser();
                stylesheetParser.Parse(binaryStylesheet);
                if (stylesheetParser.Stylesheet != null)
                {
                    stylesheetParser.Stylesheet.Apply(score);
                }

                Logger.Info(Name, "BinaryStylesheet parsed");
            }

            if (partConfiguration != null)
            {
                Logger.Info(Name, "Start Parsing Part Configuration");
                var partConfigurationParser = new PartConfigurationParser();
                partConfigurationParser.Parse(partConfiguration);
                if (partConfigurationParser.Configuration != null)
                {
                    partConfigurationParser.Configuration.Apply(score);
                }

                Logger.Info(Name, "Part Configuration parsed");
            }

            return score;
        }
    }
}

File: packages/alphatab/src/importer/GpxImporter.ts
import { BinaryStylesheet } from '@coderline/alphatab/importer/BinaryStylesheet';
import { GpifParser } from '@coderline/alphatab/importer/GpifParser';
import { GpxFileSystem } from '@coderline/alphatab/importer/GpxFileSystem';
import { PartConfiguration } from '@coderline/alphatab/importer/PartConfiguration';
import { ScoreImporter } from '@coderline/alphatab/importer/ScoreImporter';
import type { Score } from '@coderline/alphatab/model/Score';
import { Logger } from '@coderline/alphatab/Logger';
import { UnsupportedFormatError } from '@coderline/alphatab/importer/UnsupportedFormatError';
import { IOHelper } from '@coderline/alphatab/io/IOHelper';
import { LayoutConfiguration } from '@coderline/alphatab/importer/LayoutConfiguration';

/**
 * This ScoreImporter can read Guitar Pro 6 (gpx) files.
 * @internal
 */
export class GpxImporter extends ScoreImporter {
    public get name(): string {
        return 'Guitar Pro 6';
    }

    public readScore(): Score {
        // at first we need to load the binary file system
        // from the GPX container
        Logger.debug(this.name, 'Loading GPX filesystem');
        const fileSystem: GpxFileSystem = new GpxFileSystem();
        fileSystem.fileFilter = s => {
            return (
                s.endsWith('score.gpif') ||
                s.endsWith('BinaryStylesheet') ||
                s.endsWith('PartConfiguration') ||
                s.endsWith('LayoutConfiguration')
            );
        };
        fileSystem.load(this.data);
        Logger.debug(this.name, 'GPX filesystem loaded');

        let xml: string | null = null;
        let binaryStylesheetData: Uint8Array | null = null;
        let partConfigurationData: Uint8Array | null = null;
        let layoutConfigurationData: Uint8Array | null = null;
        for (const entry of fileSystem.files) {
            switch (entry.fileName) {
                case 'score.gpif':
                    xml = IOHelper.toString(entry.data!, this.settings.importer.encoding);
                    break;
                case 'BinaryStylesheet':
                    binaryStylesheetData = entry.data;
                    break;
                case 'PartConfiguration':
                    partConfigurationData = entry.data;
                    break;
                case 'LayoutConfiguration':
                    layoutConfigurationData = entry.data;
                    break;
            }
        }

        if (!xml) {
            throw new UnsupportedFormatError('No score.gpif found in GPX');
        }

        // the score.gpif file within this filesystem stores
        // the score information as XML we need to parse.
        Logger.debug(this.name, 'Start Parsing score.gpif');
        const gpifParser: GpifParser = new GpifParser();
        gpifParser.parseXml(xml, this.settings);
        Logger.debug(this.name, 'score.gpif parsed');
        const score: Score = gpifParser.score;

        if (binaryStylesheetData) {
            Logger.debug(this.name, 'Start Parsing BinaryStylesheet');
            const binaryStylesheet: BinaryStylesheet = new BinaryStylesheet(binaryStylesheetData);
            binaryStylesheet.apply(score);
            Logger.debug(this.name, 'BinaryStylesheet parsed');
        }

        let partConfigurationParser: PartConfiguration | null = null;
        if (partConfigurationData) {
            Logger.debug(this.name, 'Start Parsing Part Configuration');
            partConfigurationParser = new PartConfiguration(partConfigurationData);
            partConfigurationParser.apply(score);
            Logger.debug(this.name, 'Part Configuration parsed');
        }

        if (layoutConfigurationData && partConfigurationParser != null) {
            Logger.debug(this.name, 'Start Parsing Layout Configuration');
            const layoutConfigurationParser: LayoutConfiguration = new LayoutConfiguration(
                partConfigurationParser,
                layoutConfigurationData
            );
            layoutConfigurationParser.apply(score);
            Logger.debug(this.name, 'Layout Configuration parsed');
        }

        return score;
    }
}
